include/storage/io.hpp

```diff
@@ -192,7 +199,11 @@ class FileWriter
     {
         output_stream.open(filepath, std::ios::binary);
         if (!output_stream)
-            throw util::exception("Error opening " + filepath.string());
+        {
+            throw util::FileWriteException(std::string(filepath.string()) + ": " +
+                                               std::string(std::strerror(errno)),
```

Comment:
here and everywhere in this changeset - I don't think streams give you the guarantee to populate C's global `errno` var